### PR TITLE
Implement HPET driver

### PR DESCRIPTION
**Pull request by Thog**
This PR adds support for the "High Precision Event Timer" device.

This permits us to get better timing resolution (PIC resolution was 10ms, we now have a resolution of 100ns)

As we currently don't have IO-APIC support, the HPET is setup in "legacy mapping" and use the IRQ0 and IRQ8 for timer 0 and timer 1.
This result in the RTC device not being able to send interrupt.
For this matter, the timer 1 is setup to send an interrupt every seconds to fake this until we get IO-APIC support.

### CODE REVIEW

File: kernel/src/devices/hpet.rs [hunk 1]
```diff
@@ -75,22 +115,182 @@ impl Debug for HpetRegister {
...
+    pub fn set_comparator_value(&self, value: u64) {
+        unsafe { (*self.inner).comparator_value_low.write((value & 0xFFFF_FFFF) as u32) };
+        unsafe { (*self.inner).comparator_value_high.write((value >> 32) as u32) };
+    }
+
+    /// Set the timer accumulator value.
+    /// NOTE: The timer MUST be in periodic mode.
+    pub fn set_accumulator_value(&self, value: u64) {
+        // We update the accumulator register two times.
+        // BODY: because we are running on i386, this cause issue on qemu.
```
Comment (roblabla) [inline kernel/src/devices/hpet.rs]: Missing a TODO?

**Comment by Thog** [comment]
Ready for reviews

**Comment by Thog** [comment]
Rebased

File: kernel/src/devices/hpet.rs [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    /// NOTE: this auto-clear.
+    pub accumulator_config, set_accumulator_config: 6;
+
+    /// Set to 1 to force a 64 bit timer to operate as 32 bit one
+    /// NOTE: this as no effect on a 32 bit timer.
+    pub is_32bit_mode, set_32bit_mode: 8;
+
+    /// Timer Interrupt Route: This indicate the routing in the I/O APIC
+    /// NOTE: If the LegacyReplacement Route bit is set, then Timers 0 and 1 will have a different routing, and this bit field has no effect for those two timers.
+    /// NOTE: If the Timer FSB Interrupt bit is set, then the interrupt will be delivered directly to the FSB, and this bit field has no effect.
```
Comment (Orycterope) [inline kernel/src/devices/hpet.rs]: Use the `# Note` markdown syntax, and line wrap

File: kernel/src/devices/hpet.rs [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+    if main_timer_opt.is_none() {
+        paging::kernel_memory::get_kernel_memory().unmap(virtual_address, PAGE_SIZE);
+        return false;
+    }
+
+    // Set the tick rate in femtoseconds
+    // Kernel needs an update frequency of 100 nanoseconds.
+    let tick_rate = 100;
+    let tick_rate_nano = tick_rate * 1000000;
+    let tick_rate_femto = tick_rate_nano * 1000000;
```
Comment (Orycterope) [inline kernel/src/devices/hpet.rs]: This is really confusing. What is a "tick rate" ? Period ? Frequency ?

Nano is 10⁻⁹, not 10⁻⁶.

The `tick_rate_nano` step is useless, jump straight to femto (15 zeros).

Use the `1_000_000` number syntax, so we can count 0s.

File: kernel/src/devices/hpet.rs [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    info!("HPET frequency: {} Hz", hpet_instance.get_frequency());
+    info!("HPET tick rate: {} fs", tick_rate_femto);
+
+    let main_timer = main_timer_opt.unwrap();
+
+    // The timer must support periodic interrupt otherwise we cannot use it!
+    if !main_timer.support_periodic_interrupt() {
```
Comment (Orycterope) [inline kernel/src/devices/hpet.rs]: you should do this check earlier, with the rest of the other checks.

File: kernel/src/devices/hpet.rs [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+        paging::kernel_memory::get_kernel_memory().unmap(virtual_address, PAGE_SIZE);
+        return false;
+    }
+
+    let kernel_frequency = tick_rate_femto / u64::from(hpet_instance.get_period());
```
Comment (Orycterope) [inline kernel/src/devices/hpet.rs]: Can you prove this is correct ?

**Comment by Orycterope** [comment]
I don't recall what was the precise reason of this PR ? Did we need a higher precision timer somewhere ? Is this a necessary step to some goal ?

File: kernel/src/devices/hpet.rs (hunk 3 above)
Comment (Thog) [reply kernel/src/devices/hpet.rs]: This was for debugging purpose, I was actually using 100 ms and not 100 ns here to check coherency with the PIT (oops)

File: kernel/src/devices/hpet.rs (hunk 5 above)
Comment (Thog) [reply kernel/src/devices/hpet.rs]: the HPET device provide a period at which the counter increments in femtoseconds (10^-15 seconds)

`tick_rate_femto` is in femtoseconds so everything is in order?

**Comment by Thog** [comment]
@Orycterope This is a preparation to get the LAPIC and IO-APIC implemented (and then multicore). The final goal of this is to have multiple timer on multiple core.

File: kernel/src/devices/hpet.rs [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+    main_timer.set_comparator_value(kernel_frequency);
+
+    // Fake the RTC as in legacy mode it's taking it's IRQ.
+    // TODO: remove this when IO-APIC will be implemented
+    // Cannot fail, by spec there is at least 3 comparators defined.
+    let rtc_timer = hpet_instance.get_timer(1).unwrap();
+    if !rtc_timer.support_periodic_interrupt() {
+        warn!("RTC irq cannot be faked, clock will be broken!");
+    } else {
+        let rtc_frequency = hpet_instance.get_frequency();
```
Comment (Orycterope) [inline kernel/src/devices/hpet.rs]:
```suggestion
        let rtc_period_in_ticks = hpet_instance.get_frequency();
```

File: kernel/src/devices/hpet.rs [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+    let virtual_address = paging::kernel_memory::get_kernel_memory().map_phys_region(
+        physical_mem,
+        MappingAccessRights::READABLE | MappingAccessRights::WRITABLE,
+    );
+    let hpet_mmio = virtual_address.addr() as *mut HpetRegister;
+    let hpet_instance = Hpet::new(hpet_mmio);
+
+    // First disable the hpet if it's running.
+    if hpet_instance.is_enabled() {
+        hpet_instance.disable();
```
Comment (Orycterope) [inline kernel/src/devices/hpet.rs]: set main counter to 0

**Pull request merged**